File: src-tauri/src/operation/pipeline/component.rs
use async_trait::async_trait;
use miette::Result;

use super::{Connection, Packet};
use crate::model::{InputPort, OutputPort};

pub type Generator = dyn Fn(&str, &crate::config::Config) -> Box<dyn Component + Send>;

pub struct Constructor {
    pub kind: &'static str,
    pub label: &'static str,
    pub gen: Box<Generator>,
}

pub trait Component {
    fn id(&self) -> String;
    fn kind(&self) -> &'static str;
    fn label(&self) -> &'static str;
    fn inputs(&self) -> Vec<InputPort>;
    fn outputs(&self) -> Vec<OutputPort>;

    fn spawn(&self) -> ProcessInit;
}

#[async_trait]
pub trait Process {
    async fn run(&mut self, conn: &mut Connection) -> Result<()>;
}

pub type ProcessInit =
    std::pin::Pin<Box<dyn core::future::Future<Output = Result<Box<dyn Process + Send>>> + Send>>;

#[async_trait]
pub trait DefaultProcess: Process {
    async fn default_run(&mut self, conn: &mut Connection) -> Result<()> {
        loop {
            let Some(packet) = conn.receive().await else {
                return Ok(());
            };
            let packets = self.handler(packet).await?;

            for packet in packets {
                conn.send(packet).await?;
            }
        }
    }

    async fn handler(&mut self, packet: Packet) -> Result<Vec<Packet>>;
}

#[async_trait]
pub trait PassiveProcess: Process {
    async fn passive_run(&mut self, connection: &mut Connection) -> Result<()> {
        loop {
            let packets = self.handler().await?;
            for packet in packets {
                connection.send(packet).await?;
            }
        }
    }

    async fn handler(&mut self) -> Result<Vec<Packet>>;
}

File: operation/pipeline/component.go
package pipeline

import (
	"github.com/yuniruyuni/mrdamian/config"
	"github.com/yuniruyuni/mrdamian/model"
)

type Constructor struct {
	Kind  string
	Label string
	Gen   func(id string, c *config.Config) Component
}

type Component interface {
	ID() string
	Kind() string
	Label() string
	Inputs() model.InputPorts
	Outputs() model.OutputPorts

	Run() Process
}

type Process interface {
	Run(conn *Connection)
}

type PassiveProcess struct {
	Handler func(*Packet) []*Packet
}

func (p *PassiveProcess) Run(conn *Connection) {
	for {
		packet, err := conn.Receive()
		if err != nil {
			return
		}

		packets := p.Handler(packet)
		for _, packet := range packets {
			if err := conn.Send(packet); err != nil {
				// TODO: logging
				continue
			}
		}
	}
}

type ActiveProcess struct {
	Handler func() []*Packet
}

func (p *ActiveProcess) Run(conn *Connection) {
	for {
		packets := p.Handler()
		for _, packet := range packets {
			if err := conn.Send(packet); err != nil {
				// TODO: logging
				continue
			}
		}
	}
}
